apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java

```diff
@@ -93,6 +95,12 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
             servletRequest.getDispatcherType() == DispatcherType.REQUEST &&
             !Boolean.TRUE.equals(excluded.get())) {
 
+            ServletContext servletContext = servletRequest.getServletContext();
+            if (servletContext != null) {
+                // this makes sure service name discovery also works when attaching at runtime
```

Comment:
Why didn't we do it on the first place? We can lose metrics events for this service on startup or something?

Reply:
I don't think we can loose events with that approach. Only doing name discovery on Servlet#init had the benefit that we didn't have to do it on each transaction. But the filtering based on looking up in a set if the name has already been determined for a given servlet path should be quick enough.